File: app/src/test/kotlin/com/ssbmax/architecture/RawFirestorePathLiteralGuardTest.kt
package com.ssbmax.architecture

import org.junit.Assert.assertEquals
import org.junit.Test
import java.io.File

/**
 * Phase 2 guard test (docs/plans/CrossPlatform_SSOT): a raw Firestore
 * collection/document-path literal must never reappear in `shared` or
 * `data-firebase` outside the generated `SsbContracts` file itself. This is
 * a fast, source-grep backstop to `RawFirestorePathLiteralRule` in
 * `:detekt-rules` (the ssbmax detekt ruleset) -- text-based rather than
 * AST-based, so it also catches drift if detekt is ever skipped locally.
 */
class RawFirestorePathLiteralGuardTest {

    private val projectRoot: File =
        File(System.getProperty("user.dir") ?: ".").parentFile ?: File(".")

    @Test
    fun `no raw Firestore collection or document literal outside SsbContracts`() {
        val rawLiteralCall = Regex("""\.(collection|document)\(\s*"[^"$]*"\s*\)""")
        val exemptFiles = setOf(
            "shared/src/commonMain/kotlin/com/ssbmax/shared/contracts/SsbContracts.kt"
        )

        val offenders = listOf(
            File(projectRoot, "shared/src/commonMain"),
            File(projectRoot, "data-firebase/src/commonMain")
        ).flatMap { root ->
            root.walkTopDown()
                .filter { it.isFile && it.extension == "kt" }
                .filter { file ->
                    val relativePath = file.relativeTo(projectRoot).path.replace(File.separatorChar, '/')
                    relativePath !in exemptFiles && rawLiteralCall.containsMatchIn(file.readText())
                }
                .map { it.relativeTo(projectRoot).path.replace(File.separatorChar, '/') }
        }

        assertEquals(
            "Raw Firestore path literal(s) found -- use SsbContracts.FirestorePaths instead: $offenders",
            emptyList<String>(),
            offenders
        )
    }
}

File: functions/test/rawFirestorePathLiteralGuard.test.js
/**
 * Phase 2 guard test (docs/plans/CrossPlatform_SSOT): a raw Firestore
 * collection/document-path literal must never reappear in functions/src
 * outside the generated contracts file. FirestorePaths
 * (src/generated/contracts.cjs) is the only source of truth for a path.
 *
 * Runs via Node native test runner (node --test).
 */

const test = require('node:test');
const assert = require('node:assert/strict');
const fs = require('node:fs');
const path = require('node:path');

const srcRoot = path.join(__dirname, '..', 'src');
const exemptDir = path.join(srcRoot, 'generated');

// `.collection('literal')` / `.doc('literal')` -- a bare string-literal
// argument (no template interpolation) to a Firestore collection/doc call.
const RAW_LITERAL_CALL = /\.(collection|doc)\(\s*'[^'`]*'\s*\)/;

function walk(dir) {
  return fs.readdirSync(dir, { withFileTypes: true }).flatMap((entry) => {
    const full = path.join(dir, entry.name);
    if (entry.isDirectory()) return full === exemptDir ? [] : walk(full);
    return entry.name.endsWith('.js') ? [full] : [];
  });
}

test('functions/src has zero raw literal .collection()/.doc() calls outside generated/', () => {
  const offenders = walk(srcRoot)
    .filter((file) => RAW_LITERAL_CALL.test(fs.readFileSync(file, 'utf8')))
    .map((file) => path.relative(srcRoot, file));

  assert.deepEqual(offenders, []);
});
